### PR TITLE
WCT retries fix

### PR DESCRIPTION
**Pull request by joelridden**
Manages issue where a job can infinity be in WCT.
Now 2 different checks for max retries, check both failed and killed_WCT individually if they are less than the max retries before submitting a new job.
Added Increase WCT for IM_calc.

### CODE REVIEW

File: workflow/automation/lib/MgmtDB.py [hunk 1]
```diff
@@ -183,15 +184,27 @@ def update_entries_live(
...
+                ):
+                    # The task was killed_WCT. If there have been few enough other attempts at the task make another one
+                    logger.debug(
+                        "Task hit WCT but is able to be retried. Adding new task to the db"
+                    )
+                    self._insert_task(cur, realisation_name, process)
+                    logger.debug("New task added to the db")
 
                 if entry.status == const.Status.failed.value:
                     tasks = MgmtDB.find_dependant_task(cur, entry)
```
Comment (jpa198) [inline workflow/automation/lib/MgmtDB.py:210]: Could be merged with the above failure retries branch.
Fine as is though.

File: workflow/automation/lib/MgmtDB.py [hunk 2]
```diff
@@ -183,15 +184,27 @@ def update_entries_live(
...
-                        self._insert_task(cur, realisation_name, process)
-                        logger.debug("New task added to the db")
+                    # The task was failed. If there have been few enough other attempts at the task make another one
+                    logger.debug(
+                        "Task failed but is able to be retried. Adding new task to the db"
+                    )
+                    self._insert_task(cur, realisation_name, process)
+                    logger.debug("New task added to the db")
+
+                if (
```
Comment (jpa198) [inline workflow/automation/lib/MgmtDB.py:197]: Can be elif, as the status cannot be both failed and killed_wct

**Review by jpa198 — APPROVED** [review]

File: workflow/automation/lib/MgmtDB.py [hunk 3]
```diff
@@ -424,7 +437,7 @@ def get_runnable_tasks(
                 ).fetchall()
                 runnable_tasks.extend(
                     [
-                        (*task, self.get_retries(*task))
+                        (*task, self.get_retries(*task, get_WCT=True))
```
Comment (sungeunbae) [inline workflow/automation/lib/MgmtDB.py:440]: So this effectively controls the increase of WCT if it was a WCT failure?

Comment (joelridden) [reply workflow/automation/lib/MgmtDB.py:440]: Yes as that determines the number of past retries that have hit WCT, so when that is passed into the auto submit scripts then they adjust the WCT based on the number of past retries

File: workflow/automation/lib/MgmtDB.py [hunk 4]
```diff
@@ -319,23 +331,23 @@ def add_retries(self, n_max_retries: int):
...
             ).fetchall()
 
         failure_count = {}
-        for run_name, proc_type in errored:
+        for run_name, proc_type, state in errored:
             key = "{}__{}".format(run_name, proc_type)
             if key not in failure_count.keys():
-                failure_count.update({key: 0})
-            failure_count[key] += 1
+                failure_count.update({key: {"killed_WCT": 0, "failed": 0}})
```
Comment (jpa198) [inline workflow/automation/lib/MgmtDB.py:345]: In future if we add more failure modes we can abstract the state dictionary generation out. This works for now though

File: workflow/automation/lib/MgmtDB.py [hunk 5]
```diff
@@ -319,23 +331,23 @@ def add_retries(self, n_max_retries: int):
...
+            failure_count[key][state] += 1
 
         with connect_db_ctx(self._db_file) as cur:
             for key, fail_count in failure_count.items():
-                if fail_count >= n_max_retries:
+                if any([x >= n_max_retries for x in fail_count.values()]):
```
Comment (jpa198) [inline workflow/automation/lib/MgmtDB.py:350]: Beautiful

**Review by jpa198 — APPROVED** [review]

**Review by sungeunbae — APPROVED** [review]

**Pull request merged**